### PR TITLE
Loss, NCHW support fixed

### PR DESCRIPTION
**Pull request by squarenabla**
If labels has channels first format, transform to channels last

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -5468,14 +5468,18 @@ def init_by_layer(self, layer, layer_output_template=None):
...
     :param tf.Tensor x: (B,T,...) or (T,B,...)
     :param tf.Tensor seq_lens: (B,)
     :param bool time_major:
+    :param bool batch_feature_major:
     :return: (B*T|T*B|B',...)
     :rtype: tf.Tensor
     """
+    if batch_feature_major:
+      from TFUtil import swapaxes
+      x = swapaxes(x, 1, 2)
```
Comment (albertz) [inline TFNetworkLayer.py]: This is wrong in general. You probably mean `-1` instead of `2`.
But also, you are actually swapping it already below, so the whole change here in `_flatten_or_merge` should not be needed?

File: TFNetworkLayer.py [hunk 2]
```diff
@@ -5516,9 +5520,12 @@ def init(self, output, output_with_activation=None, target=None, layer=None):
       if self.output.have_time_axis():
         self.output_seq_lens = output.get_sequence_lengths()
         time_and_batch_dims = (self.output.time_dim_axis, self.output.batch_dim_axis)
-        assert time_and_batch_dims in [(0, 1), (1, 0)], "output time-batch-dim unexpected: %s" % self.output
+        assert time_and_batch_dims in [(0, 1), (1, 0), (2, 0)], "output time-batch-dim unexpected: %s" % self.output
+        if self.output.time_dim_axis == 2:
+          self.output = self.output.copy_with_feature_dim_axis(-1)
```
Comment (albertz) [inline TFNetworkLayer.py]: If you anyway do this, then why don't you do it right in the beginning, **before** `time_and_batch_dims = (...)`? Then you don't need to change any other code, and it's much simpler. So please do that.

Comment (albertz) [reply TFNetworkLayer.py]: Also, instead of checking `time_dim_axis == 2`, check for `is_batch_feature_major`.

File: TFNetworkLayer.py (hunk 1 above)
Comment (squarenabla) [reply TFNetworkLayer.py]: You mean `x = swapaxes(x, 0, 1)`, which goes below? But it's swapping between time_dim and batch_dim.

Comment (albertz) [reply TFNetworkLayer.py]: No. I mean what I wrote, I.e. I mean `-1` instead of `2`, i.e. `swapaxes(x, 1, -1)` (which is equivalent to `swapaxes(x, 1, 2)` if `ndim == 3`.

File: TFNetworkLayer.py [hunk 3]
```diff
@@ -5515,10 +5519,13 @@ def init(self, output, output_with_activation=None, target=None, layer=None):
...
       if self.output.have_time_axis():
         self.output_seq_lens = output.get_sequence_lengths()
+        if self.output.is_batch_feature_major:
+          self.output = self.output.copy_with_feature_dim_axis(-1)
         time_and_batch_dims = (self.output.time_dim_axis, self.output.batch_dim_axis)
         assert time_and_batch_dims in [(0, 1), (1, 0)], "output time-batch-dim unexpected: %s" % self.output
         if output_with_activation and output_with_activation.act_func is tf.nn.softmax:
-          self.output_before_softmax_flat = flatten_or_merge(output_with_activation.x, self.output_seq_lens, time_major=output.is_time_major)
+          self.output_before_softmax_flat = flatten_or_merge(output_with_activation.x, self.output_seq_lens,
+                                                             time_major=output.is_time_major, batch_feature_major=output.is_batch_feature_major)
```
Comment (albertz) [inline TFNetworkLayer.py]: You don't need that.
After you do `self.output.copy_with_feature_dim_axis(-1)` above, you will not have `is_batch_feature_major` anymore. So remove this.

File: TFNetworkLayer.py [hunk 4]
```diff
@@ -5468,14 +5468,18 @@ def init_by_layer(self, layer, layer_output_template=None):
       target=layer._get_target_value(),
       layer=layer)
 
-  def _flatten_or_merge(self, x, seq_lens, time_major):
+  def _flatten_or_merge(self, x, seq_lens, time_major, batch_feature_major=False):
```
Comment (albertz) [inline TFNetworkLayer.py]: This change here is not needed. See my other comment.

File: TFNetworkLayer.py (hunk 3 above)
Comment (squarenabla) [reply TFNetworkLayer.py]: 'self.output.copy_with_feature_dim_axis(-1)' will return the right format. But  'output_with_activation.x' will be still in the old format. So it has to be converted as well.

File: TFNetworkLayer.py [hunk 5]
```diff
@@ -5468,14 +5468,18 @@ def init_by_layer(self, layer, layer_output_template=None):
...
     :param tf.Tensor x: (B,T,...) or (T,B,...)
     :param tf.Tensor seq_lens: (B,)
     :param bool time_major:
+    :param bool batch_feature_major:
     :return: (B*T|T*B|B',...)
     :rtype: tf.Tensor
     """
+    if batch_feature_major:
+      from TFUtil import swapaxes
+      x = swapaxes(x, 1, -1)
```
Comment (albertz) [inline TFNetworkLayer.py]: I think you should use `move_axis` instead here.

File: TFNetworkLayer.py (hunk 3 above)
Comment (albertz) [reply TFNetworkLayer.py]: Yes, but it should not be done only here. Also, `output_with_activation` should be consistent to `output`, always. Otherwise later you do not know anymore, because you just know what `self.output` is.
So, just do it where you do `copy_with_feature_dim_axis`. That makes the code much simpler (you don't need to change `_flatten_or_merge` and care about other problems in other code later).

File: TFNetworkLayer.py [hunk 6]
```diff
@@ -5515,10 +5519,13 @@ def init(self, output, output_with_activation=None, target=None, layer=None):
       self.loss_norm_factor = 1.0
       if self.output.have_time_axis():
         self.output_seq_lens = output.get_sequence_lengths()
+        if self.output.is_batch_feature_major:
+          self.output = self.output.copy_with_feature_dim_axis(-1)
```
Comment (albertz) [inline TFNetworkLayer.py]: `output_with_activation` should also be changed like this here.

Comment (albertz) [reply TFNetworkLayer.py]: Esp `self.output_with_activation`.

File: TFNetworkLayer.py [hunk 7]
```diff
@@ -5677,6 +5677,12 @@ def init(self, output, output_with_activation=None, target=None, layer=None):
         else:
           assert not target.beam_size
       self.output = output
+      if self.output.is_batch_feature_major:
+        self.output = self.output.copy_with_feature_dim_axis(-1)
+        if output_with_activation:
+          from TFUtil import move_axis
+          output_with_activation = OutputWithActivation(x=move_axis(output_with_activation.x, 1, -1),
+                                                        act_func=output_with_activation.act_func)
```
Comment (albertz) [inline TFNetworkLayer.py]: This is slightly inconsistent (you operate on `self.output`, i.e. the attribute, but on `output_with_activation`, i.e. not the attribute but the local var). Why don't you make it consistent? I.e. just do it in both cases on the local var (`output` and `output_with_activation`), before the assignment `self.output = output`.

Also, instead of the check `output.is_batch_feature_major`, you should check `output.feature_dim_axis is not None and output.feature_dim_axis != output.batch_ndim - 1`.
